parser/src/main/kotlin/com/linkedin/dex/parser/JUnit4Extensions.kt

```diff
@@ -5,67 +5,30 @@
 package com.linkedin.dex.parser
 
 import com.linkedin.dex.spec.ACC_INTERFACE
-import com.linkedin.dex.spec.AnnotationItem
-import com.linkedin.dex.spec.AnnotationSetItem
 import com.linkedin.dex.spec.AnnotationsDirectoryItem
 import com.linkedin.dex.spec.ClassDefItem
 import com.linkedin.dex.spec.DexFile
-import com.linkedin.dex.spec.MethodAnnotation
+import com.linkedin.dex.spec.MethodIdItem
 
 
 /**
  * Find all methods that are annotated with JUnit4's @Test annotation
  */
-fun DexFile.findJUnit4Tests(): List<String> {
-    val testAnnotationDescriptor = "Lorg/junit/Test;"
-    val matchingItems: MutableList<String> = mutableListOf()
+fun DexFile.findJUnit4Tests(): List<TestMethod> {
+    val testAnnotationName = "org.junit.Test"
+    val classesWithAnnotations = classDefs.filter(::hasAnnotations).filterNot(::isInterface)
 
-    classDefs.filter(::hasAnnotations)
-            .filterNot(::isInterface)
-            .map { Pair(it, AnnotationsDirectoryItem.create(byteBuffer, it.annotationsOff)) }
-            .map { Pair(it.first, it.second.methodAnnotations) }
-            .map { Pair(it.first, getMethodsWithAnnotation(it.second, testAnnotationDescriptor)) }
-            .filter { it.second.isNotEmpty() }
-            .map { Pair(formatClassName(it.first), it.second) }
-            .flatMap { pair -> pair.second.map { pair.first + it }.toCollection(mutableListOf()) }
-            .toCollection(matchingItems)
-
-    return matchingItems
-}
-
-private fun hasAnnotations(classDefItem: ClassDefItem): Boolean {
-    return classDefItem.annotationsOff != 0
+    return createTestMethods(classesWithAnnotations, findMethodIds())
+            .filter { it.annotationNames.contains(testAnnotationName) }
 }
 
-fun DexFile.formatClassName(classDefItem: ClassDefItem): String {
-    var className = ParseUtils.parseClassName(byteBuffer, classDefItem, typeIds, stringIds)
-    // strip off the "L" prefix
-    className = className.substring(1)
-    className = className.replace('/', '.')
-    // strip off the ";"
-    className = className.dropLast(1)
-    // the instrument command expects test class names and method names to be separated by a "#"
-    className += "#"
-
-    return className
+/**
+ * Find methodIds we care about: any method in the class which is annotated
+ */
+private fun DexFile.findMethodIds(): (ClassDefItem, AnnotationsDirectoryItem?) -> List<MethodIdItem> {
```

Comment:
I like that idea, and just tried to make the change, but currently this is an extension method on `DexFile` and needs access to the `methodIds` property on `DexFile`, so i think it needs to stay a method